File: solutions/scala/114-invert-binary-tree.scala
/**
 * Problem 114: Invert Binary Tree (LeetCode 226)
 * Difficulty: Easy
 * Language: Scala
 */
sealed trait BTree[+A]
case object Empty extends BTree[Nothing]
case class Node[A](v: A, l: BTree[A], r: BTree[A]) extends BTree[A]

def invertTree[A](root: BTree[A]): BTree[A] = root match {
  case Empty => Empty
  case Node(v, l, r) => Node(v, invertTree(r), invertTree(l))
}

File: solutions/haskell/114-invert-binary-tree.hs
{-
  Problem 114: Invert Binary Tree (LeetCode 226)
  Difficulty: Easy
  Language: Haskell
-}
data Tree a = Empty | Node a (Tree a) (Tree a)

invertTree :: Tree a -> Tree a
invertTree Empty = Empty
invertTree (Node v l r) = Node v (invertTree r) (invertTree l)
